src/search3_rpc.erl

```diff
@@ -0,0 +1,210 @@
+    Prefix= <<DbName/binary, Signature/binary>>,
+    #{prefix => Prefix, session => Session};
+construct_index_msg(#index{dbname = DbName, sig = Signature,
+        analyzer = Analyzer, session = Session}) when is_binary(Analyzer) ->
+    Prefix= <<DbName/binary, Signature/binary>>,
+    #{prefix => Prefix, session => Session, default => #{name => Analyzer}};
+construct_index_msg(#index{dbname = DbName, sig = Signature,
+        analyzer = {Analyzer}, session= Session}) ->
+    Prefix= <<DbName/binary, Signature/binary>>,
+    case construct_analyzer_spec(Analyzer) of
+        #{name := <<"perfield">>, stopwords := Stopwords} ->
+            Fields = construct_per_fields(Analyzer),
+            Default = construct_default(Analyzer, Stopwords),
+            #{prefix => Prefix, session => Session,
+                default => Default, per_field => Fields};
+        AnalyzerSpec ->
+            #{prefix => Prefix, session => Session, default => AnalyzerSpec}
+    end.
+
+construct_analyzer_spec(Analyzer) ->
+    {_, Name} = lists:keyfind(<<"name">>, 1, Analyzer),
+    Stopwords = case lists:keyfind(<<"stopwords">>, 1, Analyzer) of
+        false ->
+            [];
+        {_, Words} ->
+            Words
+    end,
+    #{name => Name, stopwords => Stopwords}.
+
+construct_per_fields(Analyzer) ->
+    Fields1 = case lists:keyfind(<<"fields">>, 1, Analyzer) of
+        false ->
+            [];
+        {_, {Fields}} ->
+            Fields
+    end,
+    MakeMapFun = fun
+        ({Field, PerFieldAnalyzer}, Map) when is_binary(PerFieldAnalyzer) ->
+            maps:put(Field, #{name => PerFieldAnalyzer}, Map);
+        ({Field, {PerFieldAnalyzer}}, Map) ->
+            AnalyzerSpec = construct_analyzer_spec(PerFieldAnalyzer),
+            maps:put(Field, AnalyzerSpec, Map)
+    end,
+    lists:foldl(MakeMapFun, #{}, Fields1).
+
+construct_default(Analyzer, Stopwords) ->
+    Default = case lists:keyfind(<<"default">>, 1, Analyzer) of
+        false ->
+            <<"standard">>;
+        {<<"default">>, Def} ->
+            Def
+    end,
+    #{name => Default, stopwords => Stopwords}.
+
+make_fields_map(Fields) when is_list(Fields) ->
+    FieldsMapFun = fun
+        ({Name, Value, {Options}}) ->
+            M1 = #{name => Name, value => fields_value(Value)},
+            Fields1 = options_list(Options),
+            M2 = maps:from_list(Fields1),
+            maps:merge(M1, M2)
+    end,
+    lists:map(FieldsMapFun, Fields).
+
+fields_value(Value) when is_binary(Value) ->
+    #{value => {string, Value}};
+fields_value(Value) when is_number(Value) ->
+    #{value => {double, Value}};
+fields_value(Value) when is_boolean(Value) ->
+    #{value => {bool, Value}}.
+
+% This function is required we need to convert <<"stored">>, <<"facet">>,
+% <<"analzyed">> into atoms for grpc.
+options_list(Options) when is_list(Options) ->
+    [{list_to_existing_atom(?b2l(Opt)), Val} || {Opt, Val} <- Options].
```

Comment:
`facet` is not mentioned as an atom anywhere in your code, so this blows up. 

Make a clause for the three defined options and a final clause to ignore anything else.